util/chunk/codec.go

```diff
@@ -189,3 +189,141 @@ func init() {
 		allNotNullBitmap[i] = 0xFF
 	}
 }
+
+// Decoder is used to:
+// 1. decode a Chunk from a byte slice.
+// How Decoder works:
+// 1. Initialization phase: Decode a whole input byte slice to Decoder.intermChk using Codec.Decode. intermChk is
+//    introduced to simplify the implementation of decode phase. This phase uses pointer operations with less CPU andF
+//    memory cost.
+// 2. Decode phase:
+//    2.1 Set the number of rows that should be decoded to a multiple of 8 greater than
+//        targetChk.RequiredRows()-targetChk.NumRows(), this can reduce the cost when copying the srcCol.nullBitMap into destCol.nullBitMap.
+//    2.2 Append srcCol.offsets to destCol.offsets when the elements is of var-length type. And further adjust the
+//        offsets according to descCol.offsets[destCol.length]-srcCol.offsets[0].
+//    2.3 Append srcCol.nullBitMap to destCol.nullBitMap.
+// 3. Go to step 1 when the input byte slice is consumed.
+type Decoder struct {
+	intermChk    *Chunk
+	colTypes     []*types.FieldType
+	remainedRows int
+}
+
+// NewDecoder creates a new Decoder object for decode a Chunk.
+func NewDecoder(chk *Chunk, colTypes []*types.FieldType) *Decoder {
+	return &Decoder{intermChk: chk, colTypes: colTypes, remainedRows: 0}
+}
+
+// Decode decodes a Chunk from Decoder, save the remained unused bytes.
+func (c *Decoder) Decode(target *Chunk) {
+	requiredRows := target.RequiredRows() - target.NumRows()
+	// Floor requiredRows to the next multiple of 8
+	requiredRows = (requiredRows + 7) >> 3 << 3
+	if requiredRows > c.remainedRows {
+		requiredRows = c.remainedRows
+	}
+	for i := 0; i < len(c.colTypes); i++ {
+		c.decodeColumn(target, i, requiredRows)
+	}
+	c.remainedRows -= requiredRows
+}
+
+// ResetAndInit resets Decoder, and use data byte slice to init it.
+func (c *Decoder) ResetAndInit(data []byte) {
+	codec := NewCodec(c.colTypes)
+	codec.DecodeToChunk(data, c.intermChk)
+	c.remainedRows = c.intermChk.NumRows()
+}
+
+// IsFinished indicate the data byte slice is consumed.
+func (c *Decoder) IsFinished() bool {
+	return c.remainedRows == 0
+}
+
+func (c *Decoder) decodeColumn(target *Chunk, ordinal int, requiredRows int) {
+	elemLen := getFixedLen(c.colTypes[ordinal])
+	numDataBytes := int64(elemLen * requiredRows)
+	srcCol := c.intermChk.columns[ordinal]
+	destCol := target.columns[ordinal]
+
+	if elemLen == varElemLen {
```

Comment:
For var-length types, we need to adjust the offsets after appending to \`descCol\`.